flink-autoscaler/src/main/java/org/apache/flink/autoscaler/IntermediateScalingResult.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.autoscaler;
+
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Class for storing intermediate scaling results. */
+public class IntermediateScalingResult {
+
+    private final Map<JobVertexID, ScalingSummary> scalingSummaries;
+    private final List<JobVertexID> bottlenecks;
+
+    private double backpropagationScaleFactor = 1.0;
+
+    public IntermediateScalingResult() {
+        scalingSummaries = new HashMap<>();
+        bottlenecks = new ArrayList<>();
+    }
+
+    void addScalingSummary(JobVertexID vertex, ScalingSummary scalingSummary) {
+        scalingSummaries.put(vertex, scalingSummary);
+    }
+
+    void addBottleneckVertex(JobVertexID bottleneck, double factor) {
+        bottlenecks.add(bottleneck);
+        backpropagationScaleFactor = Math.min(backpropagationScaleFactor, factor);
```

Comment:
The scale factor works as follows: the minimum possible value is picked and pushed to sources, lowering target capacity by this factor on each vertex. This approach works fine for most of the jobs:

1. Starting from a bottleneck vertex, the capacity of all upstream vertices of the bottleneck is reduced by the factor
2. During propagation, agation source operators are reached and their capacity is reduced
3. It affects vertices that may not be directly connected with the initial bottleneck
4. Repeating steps 2 and 3 will adjust all vertices in the connected components

I think case with 2 and more connected components (e.g. graph source1 -> op1 -> sink1; source2 -> op2 -> sink2) appears rarely.

As an alternative, bottlenecks can be iterated in bottleneck factor decreasing order, making propagation more accurate, but it takes more time for scaling (O(N^2) against O(N)) and is harder to maintain and is less predictable.

What do you think? Should we use more complex logic for propagation?

Reply:
I think we might also have to consider the output ratios when propagating the bottleneck backwards . 

So technically speaking if we want to be completely precise we can do this in a single pass if we start computing the target rates from the sinks. Once the actual scaled rate is computed we have to propagate the diff compared to the original one back based on the output ratio to the upstream tasks.

Reply:
I think Gyula has a point. The current implementation works for simple jobs but there are many jobs with more complex uncorrelated branches which would lead to unnecessary scale downs or prevent upscales (if scale down is disabled). Using the output ratios would allow us to precisely feed back the bottleneck ratios and avoid any accidental backpropagation.